// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package db;

// Interface exported by the server.
service Db {
  rpc CreateTable(stream SourceIpc) returns (Message) {}
  rpc InsertTable(stream SourceIpc) returns (Message) {}
  rpc UpsertTable(stream SourceIpc) returns (Message) {}

  // Table based handling
  rpc MaterializeTable(Table) returns (Message) {}
  rpc GetTableInfo(Table) returns (TableInfo) {}
  // rpc DropTable(Table) returns (Message) {}
  // rpc Offload(Table) returns (Message) {}

  // SwapPartition
  // DeletePartition

  rpc SelectIpc(stream Sql) returns (stream SqlResults) {}
}

message Message {
  string message = 1;
}

message Database {
  string name = 1;
}

message Table {
  string schema = 1;
  string table = 2;
}

message TableInfo {
  repeated string columns = 1;
  repeated string dtypes = 2;
  uint32 rows = 3;
  uint32 parts = 4;
}

message SourceIpc {
  string schema = 1;
  string table = 2;
  bytes data = 3;
  repeated string partitions = 4;
  repeated string buckets = 5;
  repeated string keys = 6;
}

message Sql {
  string sql = 1;
  optional uint32 qid = 2;
}

message SqlResults {
  bytes data = 1;
  uint32 rows = 2;
  repeated string columns = 3;
  repeated string dtypes = 4;
  optional uint32 qid = 5;
}